### PR TITLE
PIPE2D-576: Generate changelog for DRP releases

### PR DESCRIPTION
**Pull request by hassanxp**

### CODE REVIEW

File: python/pfs/pipe2d/git_changelog.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+                              auth,
+                              GITHUB_API_MAX_PAGES)
+
+    # Check whether for GitHub API calls, rate limit has been exceeded.
+    # If so, cannot cannot continue.
+    r = mediator.getRateLimit()
+    remaining_requests = r['resources']['core']['remaining']
+    logger.debug(f'There are {remaining_requests} '
+                 'github API requests remaining.')
+    if remaining_requests < 1:
```
[the remark this answers is not in the document]
Comment (hassanxp) [reply python/pfs/pipe2d/git_changelog.py:484]: About 11.

File: python/pfs/pipe2d/git_changelog.py [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+
+        Returns
+        -------
+        `dict`
+            The GitHub response. The structure
+            depends on the specific GitHub API
+            request being made.
+        """
+        r = requests.get(url, params=self.params, auth=self.auth)
+        return json.loads(r.text)
```
Comment (PaulPrice) [inline python/pfs/pipe2d/git_changelog.py:131]: Can do:

    return requests.get(url, params=self.params, auth=self.auth).json()

File: python/pfs/pipe2d/git_changelog.py [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+(that is, an error message is raised and the program exits)
+"""
+
+GITHUB_API_TICKET_REGEX = r'Subaru-PFS:tickets/([A-Z0-9]+-[0-9]+)'
+"""JIRA ticket regular expression (`str`).
+The JIRA ticket name is stored in the GitHub API in the above
+format. This is used to parse the ticket number.
+"""
+
+TAG_REGEX = r"([0-9]+)\.([0-9]+)(\.[0-9]+)?"
```
[the remark this answers is not in the document]
Comment (hassanxp) [reply python/pfs/pipe2d/git_changelog.py]: Agreed. Done.

File: python/pfs/pipe2d/git_changelog.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+
+class Paginator:
+    """ Combines GitHub API REST v3 requests
+    that are provided across multiple pages.
+
+    Typically a call through the GitHub API
+    will paginate the requested items
+    to keep their servers happy.
+
+    See also
```
[the remark this answers is not in the document]
Comment (hassanxp) [reply python/pfs/pipe2d/git_changelog.py:82]: Done.

File: python/pfs/pipe2d/gitChangelog.py [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+    """Extracts summary of ticket from an external source (eg JIRA).
+
+    Parameters
+    ----------
+    ticket : `str`
+        Ticket identifier.
+
+    Returns
+    -------
+    `str`
```
Comment (PaulPrice) [inline python/pfs/pipe2d/gitChangelog.py:346]: Needs name in addition to type.

File: python/pfs/pipe2d/gitChangelog.py [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+    Converts a tagname ("m.n" or "m.n.p") into a key for sorting.
+
+    Parameters
+    ----------
+    tagname : `str`
+        name of release tag
+
+    Returns
+    -------
+    `int`
```
Comment (PaulPrice) [inline python/pfs/pipe2d/gitChangelog.py:392]: Need name in addition to type.

File: python/pfs/pipe2d/gitChangelog.py [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+                        choices=['trace', 'debug', 'info',
+                                 'warn', 'error', 'fatal'],
+                        help=("logging level"))
+
+    args = parser.parse_args()
+    process(args)
+
+
+if __name__ == "__main__":
+    main()
```
Comment (PaulPrice) [inline python/pfs/pipe2d/gitChangelog.py:548]: Remove these lines.

File: python/pfs/pipe2d/gitChangelog.py [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+        ------
+        response : `list` [`dict`]
+            The GitHub response. Each element in the list
+            corresponds to a separate page. The structure
+            of each element depends on the GitHub API
+            call.
+        """
+        for _ in range(self.max_pages):
+            r = self._getRequest(url)
+            yield json.loads(r.text)
```
[the remark this answers is not in the document]
Comment (hassanxp) [reply python/pfs/pipe2d/gitChangelog.py:153]: The header information in the Response, `r`, is needed in line 154.

**Pull request merged**